src/computation/1_computation_parallel.cpp

```diff
@@ -0,0 +1,379 @@
+
+        // step 7: calculate the final velocities
+        computeVelocities();
+
+        std::cout << "Computed velocities" << std::endl;
+
+        // step 9: write output
+        if (std:floor(currentTime + dt) == sec +1.)
+        {
+            outputWriterParaview_->writeFile(currentTime);
+            outputWriterText_->writeFile(currentTime);
+            sec = sec +1;
+        }
+
+    }
+}
+
+void ComputationParallel::computeTimeStepWidth()
+{
+    double boundary_diffusion = 0.;
+
+    // boundary from diffusion
+    if (meshWidth_[0] == meshWidth_[1])
+    {
+        boundary_diffusion = (settings_.re * meshWidth_[0] * meshWidth_[1]) / 4.;
+    } else
+    {
+        double h2x = meshWidth_[0] * meshWidth_[0];
+        double h2y = meshWidth_[1] * meshWidth_[1];
+        boundary_diffusion = (settings_.re / 2.) * (h2x * h2y) * (1./(h2x + h2y));
+      
+        std::cout << settings_.re << " " << h2x << " " << h2y << std::endl;
+    }
+
+    // calculate max absolute velocities
+    double u_max = 0.;
+    for ( int j = discretization_->uJBegin(); j < discretization_->uJEnd(); j++)
+    { 
+        for ( int i = discretization_->uIBegin(); i < discretization_->uIEnd(); i++)
+        {
+            double value = std::abs(discretization_->u(i,j));
+            if (value > u_max)
+            {
+                u_max = value;
+            }
+        }
+    }
+
+    double v_max = 0.;
+    for ( int j = discretization_->vJBegin(); j < discretization_->vJEnd(); j++)
+    { 
+        for ( int i = discretization_->vIBegin(); i < discretization_->vIEnd(); i++)
+        {
+            double value = std::abs(discretization_->v(i,j));
+            if (value > v_max)
+            {
+                v_max = value;
+            }
+        }
+    }
+
+    // boundary from convection
+    double boundary_convection_u = meshWidth_[0] / u_max;
+    double boundary_convection_v = meshWidth_[1] / v_max;
+
+    // together
+    double min_dt = std::min({boundary_diffusion, boundary_convection_u, boundary_convection_v});
+
+    std::cout << "dt boundaries - diffusion: " << boundary_diffusion << " convection_u: " << boundary_convection_u << " convection_v: " << boundary_convection_v << std::endl;
+
+    // security factor
+    double dt_local =  std::min(min_dt * settings_.tau, settings_.maximumDt);
+
+    // minimize by communication with the other processes
+    double dt_global;
+    MPI_Allreduce(&dt_local, &dt_global, 1, MPI_DOUBLE, MPI_MIN, MPI_COMM_WORLD);
+    
+    dt_ = dt_global;
+}
+    
+void ComputationParallel::applyBoundaryValues()
+{
+    // first horizontal pairwise exchange
+    // the even processes: left, then right
+    if ((partitioning.ownRankNo() % 2) == 0)
+    {
+        if (partitioning_.ownPartitionContainsLeftBoundary())
+        {
+            applyBoundaryValuesLeft();
+        } else 
+        {
+            // uExchangeVertical
+            // vExchangeVertical
+            // pExchangeVertical
```

Comment:
Pressure Exchange only in Pressure Silver itself

Reply:
yes